include/real/exact_number.hpp

```diff
@@ -643,46 +635,122 @@ namespace boost {
-                    integer_part = integer_part.substr(0, i + 1);
-                }
-                if (integer_part.empty()) {
-                    i = 0;
-                    while (decimal_part[i] == '0' && i < decimal_part.length()) {
-                        ++i;
-                        --exponent;
+
+                for (; index < number.size(); index++) {
+                    // we should not have any characters in the input except 'e' and '.', and
+                    // '.' can only come before e, and these characters can only occur once
+                    if (!on_exponent) {
+                        if(number[index] == 'e') {
+                            on_exponent = true;
+
+                            if (number[index+1] == '-') {
+                                exp_positive = false;
+                                index++;
+                            } else if (number[index+1] == '+') {
+                                index++;
+                            }
+
+                            continue;
+                        }
+
+                        if (!on_decimal) {
+                            if(number[index] == '.') {
+                                on_decimal = true;
+                                decimal_index = index;
+                                continue;
+                            }
+                        }
+                    }
+                    if (!std::isdigit(number[index])) {
+                        throw boost::real::invalid_string_number_exception();
+                    } else { // number[index] is a digit
+                        if (on_exponent) {
+                            exponent *= 10;
+                            exponent += number[index] - '0';
+                            if (exponent < 0) {
+                                throw exponent_overflow_exception();
+                            }
+                        } else if (on_decimal) {
+                            decimal_part.push_back(number[index]);
+                        } else { // we're on the integral part
+                            integer_part.push_back(number[index]);
+                        }
                     }
-                    decimal_part = decimal_part.substr(i);
                 }
+                if (!exp_positive)
+                    exponent *= -1;
+
+                // remove lhs zeros of integer_part
+                auto first_nonzero_it = std::find_if(integer_part.begin(), integer_part.end(), 
+                                        [](char r) { return r != '0';}
+                                        );
+                integer_part.erase(integer_part.begin(), first_nonzero_it);
+
+                if (integer_part.empty() && decimal_part.empty()) {
+                    return {integer_part, decimal_part, 0, true}; // shouldn't matter if 0 is  + or -
+                } else {
+                    if (decimal_part.empty()) { // we have an integer. remove rhs zeros, and add to exponent.
```

Comment:
There is a lot of "find_if" and iterators around. I'm pretty sure most of this can be solved in an iteration with cases. 
This is basically a FSM problem. I'm sure that if you write down the model, you can  easily translate it into a single iteration with cases. The only thing that does not fit in the FSM is the right hand sequence of zeros, but thats easily solved by accumulating sequences of zeros each time you see one in a second variable and not adding them to the original number right away, until you find a digit different than zero, if you find one, you immediately add those you accumulated and reset the accumulator, else, if you reach the end you have the count of digits for the exponent ready.
However, don't worry about this right now.